# [CSC8101] Big Data Analytics - 2022 Spark Coursework

## Coursework overview

### Inputs

- **NYC Taxi Trips dataset** - list of recorded taxi trips, each with several characteristics, namely: distance, number of passengers, origin zone, destination zone and trip cost (total amount charged to customer).
- **NYC Zones dataset** - list of zones wherein trips can originate/terminate.

### Tasks

1. Data cleaning
  1. Remove "0 distance" and 'no passengers' records.
  2. Remove outlier records. 
2. Add new columns
  1. Join with zones dataset
  2. Compute the unit profitability of each trip
3. Zone summarisation and ranking
  1. Summarise trip data per zone
  2. Obtain the top 10 ranks according to:
    1. The total trip volume
    2. Their average profitabilitiy
    3. The total passenger volume
4. Record the total and task-specific execution times for each dataset size and format.

### How to

###### Code structure and implementation

- You must implement your solution to each task in the provided function code skeleton.
- The task-specific functions are combined together to form the full pipeline code, executed last (do not modify this code).
- Before implementing the specified function skeleton, you should develop and test your solution on separate code cells (create and destroy cells as needed).

###### Development

- Develop an initial working solution for the 'S' dataset and only then optimise it for larger dataset sizes.
- To perform vectorised operations on a DataFrame:
  - use the API docs to look for existing vectorised functions in: https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql.html#functions
  - if a customised function is required (e.g. to add a new column based on a linear combination of other columns), implement your own User Defined Function (UDF). See:  https://spark.apache.org/docs/latest/sql-ref-functions-udf-scalar.html
- Use only the `pyspark.sql` API - documentation link below - (note that searching through the docs returns results from the `pyspark.sql` API together with the `pyspark.pandas` API):
  - https://spark.apache.org/docs/3.2.0/api/python/reference/pyspark.sql.html
- Periodically download your notebook to your computer as backup and safety measure against accidental file deletion.
 
###### Execution time measurement

- Execution time is calculated and returned by the Spark Engine and shown in the output region of the cell.
- To measure the execution time of a task you must perform a `collect` or similar operation (e.g. `take`) on the returned DataFrame.

## Task 0 - Read data

The code below is ready to run. **Do not modify this code**. It does the following:

- Reads the 'zones' dataset into variable 'zone_names'
- Defines the `init_trips` function that allows you to read the 'trips' dataset (from the DBFS FileStore) given the dataset size ('S' to 'XXL') and format ('parquet' or 'delta') as function arguments
- Defines the `pipeline` function, called in Task 4 to measure the execution time of the entire data processing pipeline
- Shows you how to call the `init_trips` function and display dataset characteristics (number of rows, schema)

In [0]:
## global imports
import pyspark.sql as ps
import pyspark.sql.functions as pf
import pandas as pd

# Load zone names dataset - (much faster to read small file from git than dbfs)
zones_file_url = 'https://raw.githubusercontent.com/NewcastleComputingScience/csc8101-coursework/main/02-assignment-spark/taxi_zone_names.csv'
zone_names = spark.createDataFrame(pd.read_csv(zones_file_url))

# Function to load trips dataset by selected dataset size
def init_trips(size = 'S', data_format = "parquet", taxi_folder = "/FileStore/tables/taxi"):     
    
    files = {
        'S'  : ['2021_07'],
        'M'  : ['2021'],
        'L'  : ['2020_21'],
        'XL' : ['1_6_2019', '7_12_2019'],
        'XXL': ['1_6_2019', '7_12_2019', '2020_21']
    }
    
    # validate input dataset size
    if size not in files.keys():
        print("Invalid input dataset size. Must be one of {}".format(list(files.keys())))
        return None               
    
    if data_format == "parquet":
        filenames = list(map(lambda s: f'{taxi_folder}/parquet/tripdata_{s}.parquet', files[size]))
        trips_df = spark.read.parquet(filenames[0])
        
        for name in filenames[1:]:
            trips_df = trips_df.union(spark.read.parquet(name))
            
    elif data_format == "delta":
        filenames = f"{taxi_folder}/delta/taxi-{size}-delta/"
        trips_df = spark.read.format("delta").load(filenames)
    
    else:
        print("Invalid data format. Must be one of {}".format(['parquet', 'delta']))
        return None
        
    print(
    """
    Trips dataset loaded!
    ---
      Size: {s}
      Format: {f}
      Tables loaded: {ds}
      Number of trips (dataset rows): {tc:,}
    """.format(s = size, f = data_format, ds = filenames, tc = trips_df.count()))
    
    return trips_df

# helper function to print dataset row count
def print_count(df):
    print("Row count: {t:,}".format(t = df.count()))

def pipeline(trips_df, with_task_12 = False, zones_df = zone_names):
    # Do not edit
    #---

    ## Task 1.1
    _trips_11 = t11_remove_zeros(trips_df)

    ## Task 1.2
    if with_task_12:
        _trips_12 = t12_remove_outliers(_trips_11)
    else:
        _trips_12 = _trips_11

    ## Task 2.1
    _trips_21 = t21_join_zones(_trips_12, zones_df = zone_names)

    ## Task 2.2
    _trips_22 = t22_calc_profit(_trips_21)

    ## Task 3.1
    _graph = t31_summarise_trips(_trips_22)

    ## Task 3.2
    _zones = t32_summarise_zones_pairs(_graph)

    _top10_trips     = t32_top10_trips(_zones)
    _top10_profit    = t32_top10_profit(_zones)
    _top10_passenger = t32_top10_passenger(_zones)
    
    return([_top10_trips, _top10_profit, _top10_passenger])

In [0]:
SIZE = 'S'
DATA_FORMAT = 'parquet'

# Load trips dataset
trips = init_trips(SIZE, DATA_FORMAT)

# uncomment line only for small datasets
# trips.take(1)


    Trips dataset loaded!
    ---
      Size: S
      Format: parquet
      Tables loaded: ['/FileStore/tables/taxi/parquet/tripdata_2021_07.parquet']
      Number of trips (dataset rows): 2,898,033
    


In [0]:
print_count(trips)

Row count: 2,898,033


In [0]:
display(trips).take(5)

index VendorID tpep_pickup_datetime tpep_dropoff_datetime passenger_count trip_distance RatecodeID store_and_fwd_flag PULocationID DOLocationID payment_type fare_amount extra mta_tax tip_amount tolls_amount improvement_surcharge total_amount congestion_surcharge cab_type lpep_pickup_datetime lpep_dropoff_datetime ehail_fee trip_type 0 1.0 2021-07-01 00:08:51 2021-07-01 00:13:05 1.0 0.8 1.0 N 90 68 1.0 5.0 3.0 0.5 0.0 0.0 0.3 8.8 2.5 Y null null null null 1 1.0 2021-07-01 00:22:39 2021-07-01 00:25:58 1.0 0.9 1.0 N 113 90 2.0 5.0 3.0 0.5 0.0 0.0 0.3 8.8 2.5 Y null null null null 2 1.0 2021-07-01 00:48:33 2021-07-01 00:54:58 1.0 2.8 1.0 N 88 232 2.0 10.0 3.0 0.5 0.0 0.0 0.3 13.8 2.5 Y null null null null 3 1.0 2021-07-01 00:59:44 2021-07-01 01:07:09 1.0 1.4 1.0 N 79 249 1.0 7.0 3.0 0.5 1.5 0.0 0.3 12.3 2.5 Y null null null null 4 1.0 2021-07-01 00:08:35 2021-07-01 00:16:28 0.0 2.0 1.0 N 142 238 1.0 8.5 3.0 0.5 0.0 0.0 0.3 12.3 2.5 Y null null null null 5 1.0 2021-07-01 00:10:49 2021-07-01 00:18:42 1.0 1.6 1.0 N 114 90 1.0 7.5 3.0 0.5 1.5 0.0 0.3 12.8 2.5 Y null null null null 6 1.0 2021-07-01 00:28:56 2021-07-01 00:36:11 1.0 1.8 1.0 N 90 144 1.0 7.5 3.0 0.5 2.0 0.0 0.3 13.3 2.5 Y null null null null 7 1.0 2021-07-01 00:45:10 2021-07-01 00:52:46 1.0 2.0 1.0 N 114 48 1.0 8.5 3.0 0.5 2.45 0.0 0.3 14.75 2.5 Y null null null null 8 1.0 2021-07-01 00:55:43 2021-07-01 01:10:56 1.0 5.7 1.0 N 48 152 2.0 18.5 3.0 0.5 0.0 0.0 0.3 22.3 2.5 Y null null null null 9 1.0 2021-07-01 00:04:24 2021-07-01 00:13:23 1.0 1.8 1.0 N 234 148 1.0 8.5 3.0 0.5 2.45 0.0 0.3 14.75 2.5 Y null null null null 10 1.0 2021-07-01 00:17:54 2021-07-01 00:27:11 2.0 3.5 1.0 N 79 141 1.0 11.5 3.0 0.5 2.0 0.0 0.3 17.3 2.5 Y null null null null 11 1.0 2021-07-01 00:48:52 2021-07-01 01:01:32 1.0 7.5 1.0 N 138 69 1.0 22.0 1.75 0.5 6.2 6.55 0.3 37.3 0.0 Y null null null null 12 2.0 2021-07-01 00:30:31 2021-07-01 00:37:19 1.0 1.02 1.0 N 230 164 1.0 6.5 0.5 0.5 0.0 0.0 0.3 10.3 2.5 Y null null null null 13 2.0 2021-07-01 00:48:57 2021-07-01 00:58:47 1.0 1.15 1.0 N 249 79 1.0 8.0 0.5 0.5 2.36 0.0 0.3 14.16 2.5 Y null null null null 14 2.0 2021-07-01 00:04:17 2021-07-01 00:12:49 1.0 2.16 1.0 N 48 90 1.0 8.5 0.5 0.5 2.0 0.0 0.3 14.3 2.5 Y null null null null 15 2.0 2021-07-01 00:16:31 2021-07-01 00:19:08 1.0 1.07 1.0 N 90 48 1.0 5.0 0.5 0.5 0.75 0.0 0.3 9.55 2.5 Y null null null null 16 2.0 2021-07-01 00:27:51 2021-07-01 00:32:00 1.0 1.56 1.0 N 142 238 1.0 6.5 0.5 0.5 2.06 0.0 0.3 12.36 2.5 Y null null null null 17 1.0 2021-07-01 00:15:12 2021-07-01 00:31:56 3.0 8.5 1.0 Y 132 98 2.0 24.5 1.75 0.5 0.0 0.0 0.3 27.05 0.0 Y null null null null 18 1.0 2021-07-01 00:57:47 2021-07-01 01:13:02 3.0 8.4 1.0 Y 132 222 2.0 25.0 1.75 0.5 0.0 0.0 0.3 27.55 0.0 Y null null null null 19 2.0 2021-07-01 00:49:52 2021-07-01 01:13:59 1.0 17.07 2.0 N 132 74 1.0 52.0 0.0 0.5 10.0 6.55 0.3 70.6 0.0 Y null null null null 20 1.0 2021-07-01 00:54:28 2021-07-01 00:57:57 1.0 0.8 1.0 N 48 161 1.0 5.0 3.0 0.5 2.2 0.0 0.3 11.0 2.5 Y null null null null 21 1.0 2021-07-01 00:13:15 2021-07-01 00:31:10 1.0 5.1 1.0 N 107 179 1.0 17.5 3.0 0.5 4.25 0.0 0.3 25.55 2.5 Y null null null null 22 2.0 2021-07-01 17:50:17 2021-07-01 17:59:21 1.0 2.34 1.0 N 13 246 1.0 9.5 0.5 0.5 2.66 0.0 0.3 15.96 2.5 Y null null null null 23 2.0 2021-07-01 18:05:35 2021-07-01 18:39:18 1.0 5.79 1.0 N 68 232 1.0 23.5 0.5 0.5 6.82 0.0 0.3 34.12 2.5 Y null null null null 24 1.0 2021-07-01 00:23:15 2021-07-01 00:45:06 1.0 10.0 1.0 N 138 239 1.0 29.5 4.25 0.5 8.2 6.55 0.3 49.3 2.5 Y null null null null 25 1.0 2021-07-01 00:51:07 2021-07-01 00:53:55 1.0 0.6 1.0 N 142 239 1.0 4.5 3.0 0.5 1.65 0.0 0.3 9.95 2.5 Y null null null null 26 1.0 2021-07-01 00:09:20 2021-07-01 00:21:14 1.0 4.0 1.0 N 107 87 1.0 14.5 3.0 0.5 0.0 0.0 0.3 18.3 2.5 Y null null null null 27 1.0 2021-07-01 00:30:51 2021-07-01 00:45:40 1.0 5.0 1.0 N 231 142 1.0 16.5 3.0 0.5 4.05 0.0 0.3 24.35 2.5 Y null null null null 28 1.0 2021-07-01 00:08:19 2021-07-01 00:47:23 1.0 18.6 2.0 N 132

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<command-2112130307445588> in <cell line: 1>()
----> 1 display(trips).take(5)

AttributeError: 'NoneType' object has no attribute 'take'

In [0]:
# dataset schemas
trips.printSchema()

root
 |-- index: long (nullable = true)
 |-- VendorID: double (nullable = true)
 |-- tpep_pickup_datetime: string (nullable = true)
 |-- tpep_dropoff_datetime: string (nullable = true)
 |-- passenger_count: double (nullable = true)
 |-- trip_distance: double (nullable = true)
 |-- RatecodeID: double (nullable = true)
 |-- store_and_fwd_flag: string (nullable = true)
 |-- PULocationID: long (nullable = true)
 |-- DOLocationID: long (nullable = true)
 |-- payment_type: double (nullable = true)
 |-- fare_amount: double (nullable = true)
 |-- extra: double (nullable = true)
 |-- mta_tax: double (nullable = true)
 |-- tip_amount: double (nullable = true)
 |-- tolls_amount: double (nullable = true)
 |-- improvement_surcharge: double (nullable = true)
 |-- total_amount: double (nullable = true)
 |-- congestion_surcharge: double (nullable = true)
 |-- cab_type: string (nullable = true)
 |-- lpep_pickup_datetime: string (nullable = true)
 |-- lpep_dropoff_datetime: string (nullable = true)
 |--

In [0]:
display(trips[['PULocationID', 'DOLocationID', 'trip_distance', 'passenger_count', 'total_amount']].take(5))

PULocationID DOLocationID trip_distance passenger_count total_amount 90 68 0.8 1.0 8.8 113 90 0.9 1.0 8.8 88 232 2.8 1.0 13.8 79 249 1.4 1.0 12.3 142 238 2.0 0.0 12.3

In [0]:
zone_names.printSchema()

root
 |-- LocationID: long (nullable = true)
 |-- Borough: string (nullable = true)
 |-- Zone: string (nullable = true)
 |-- service_zone: string (nullable = true)



In [0]:
display(zone_names.take(5))

LocationID Borough Zone service_zone 1 EWR Newark Airport EWR 2 Queens Jamaica Bay Boro Zone 3 Bronx Allerton/Pelham Gardens Boro Zone 4 Manhattan Alphabet City Yellow Zone 5 Staten Island Arden Heights Boro Zone

## Task 1 - Filter rows

**Input:** trips dataset

### Task 1.1 - Remove "0 distance" and 'no passengers' records

Remove dataset rows that represent invalid trips:

- Trips where `trip_distance == 0` (no distance travelled)
- Trips where `passenger_count == 0` and `total_amount == 0` (we want to retain records where `total_amount` > 0 - these may be significant as the taxi may have carried some parcel, for example)

Altogether, a record is removed if it satisfies the following conditions:

`trip_distance == 0` or `(passenger_count == 0` and `total_amount == 0)`.

**Recommended:** Select only the relevant dataset columns for this and subsequent tasks: `['PULocationID', 'DOLocationID', 'trip_distance', 'passenger_count', 'total_amount')]`

### Task 1.2 - Remove outliers using the modified z-score

Despite having removed spurious "zero passengers" trips in task 1.1, columns `total_amount` and `trip_distance` contain additional outlier values that must be identified and removed.

To identify and remove outliers, you will use the modified [z-score](https://en.wikipedia.org/wiki/Standard_score) method.
The modified z-score uses the median and [Median Absolute Deviation](https://en.wikipedia.org/wiki/Median_absolute_deviation) (MAD), instead of the mean and standard deviation, to determine how far an observation (indexed by i) is from the mean:

$$z_i = \frac{x_i - \mathit{median}(\mathbf{x})}{\mathbf{MAD}},$$

where x represents the input vector, xi is an element of x and zi is its corresponding z-score. In turn, the MAD formula is:

$$\mathbf{MAD} = 1.483 * \mathit{median}(\big\lvert x_i - \mathit{median}(\mathbf{x})\big\rvert).$$

Observations with **high** (absolute) z-score are considered outlier observations. A score is considered **high** if its __absolute z-score__ is larger than a threshold T = 3.5:

$$\big\lvert z_i \big\rvert > 3.5.$$

where T represents the number of unit standard deviations beyond which a score is considered an outlier ([wiki](https://en.wikipedia.org/wiki/68%E2%80%9395%E2%80%9399.7_rule)).

This process is repeated twice, once for each of the columns `total_amount` and `trip_distance` (in any order).

**Important:** Use the surrogate function [`percentile_approx`](https://spark.apache.org/docs/3.2.0/api/python/reference/api/pyspark.sql.functions.percentile_approx.html?highlight=percentile#pyspark.sql.functions.percentile_approx) to estimate the median (calculating the median values for a column is expensive as it cannot be parallelised efficiently).

In [0]:


def t11_remove_zeros(df):
    # input: trips dataset
    df=df.filter(~(df.trip_distance ==0))
    df=df.filter(~((df.passenger_count == 0) & (df.total_amount ==0)))
    df=df.select(['PULocationID', 'DOLocationID', 'trip_distance', 'passenger_count', 'total_amount'])
    # THE query will allow (passenger_count=0,total_amount>0) - delivery, (passenger_count>0,total_amount=0)-I dont know but not (passenger_count=0,total_amount=0)-invalid
    return df

In [0]:
# execute task 1.1
trips_11 = t11_remove_zeros(trips)

print_count(trips_11)
# display(trips_11.take(10))

Row count: 2,858,164


In [0]:


def t12_remove_outliers(df):
  median_total_amount=df.agg(pf.percentile_approx("total_amount", 0.5, pf.lit(1000))).collect()[0][0]
  median_trip_distance=df.agg(pf.percentile_approx("trip_distance", 0.5, pf.lit(1000))).collect()[0][0]
  delta_median_total_amount_abs=pf.abs(df.total_amount-median_total_amount)
  delta_median_trip_distance_abs=pf.abs(df.trip_distance-median_trip_distance)
  mad_total_amount=df.agg(pf.percentile_approx(delta_median_total_amount_abs, 0.5, pf.lit(1000))).collect()[0][0]*1.483
  mad_trip_distance=df.agg(pf.percentile_approx(delta_median_trip_distance_abs, 0.5, pf.lit(1000))).collect()[0][0]*1.483
  z_score_total_amount=((df.total_amount-median_total_amount)/mad_total_amount)
  z_score_trip_distance=((df.trip_distance-median_trip_distance)/mad_trip_distance)
  df=df.withColumn('z_score_total_amount',z_score_total_amount)
  df=df.withColumn('z_score_trip_distance',z_score_trip_distance)
  df=df.filter((pf.abs(df.z_score_total_amount)<3.5) & (pf.abs(df.z_score_trip_distance)<3.5))
  df=df.select(['PULocationID', 'DOLocationID', 'trip_distance', 'passenger_count', 'total_amount'])
  return df

In [0]:
# execute task 1.2
trips_12 = t12_remove_outliers(trips_11)

print_count(trips_12)
# display(trips_12.take(10))

Row count: 2,495,635


## Task 2 - Compute new columns

### Task 2.1 - Zone names

Obtain the **start** and **end** zone names of each trip by joining the `trips` and `zone_names` datasets (i.e. by using the `zone_names` dataset as lookup table).

**Note:** The columns containing the start and end zone ids of each trip are named `PULocationID` and `DOLocationID`, respectively.

### Task 2.2 - Unit profitability

Compute the column `unit_profitabilty = total_amount / trip_distance`.

In [0]:
# develop your solution here (create/destroy cells as needed) and then implement it in the functions below

In [0]:
# Your solution implementation to task 2.1 goes HERE
def t21_join_zones(df, zones_df = zone_names):
    # input: output of task 1.2 and zone_names dataset
    k=df.alias('a').join(zones_df,(df.PULocationID==zones_df.LocationID),'left').select('a.*',zones_df.Zone.alias("start_zone"))
    df=k.alias('c').join(zones_df.alias("f"),(k.DOLocationID==zones_df.LocationID),'left').select('c.*',pf.col('f.Zone').alias("end_zone"))
    return df

In [0]:
# execute task 2.1
trips_21 = t21_join_zones(trips_12, zones_df = zone_names)

print_count(trips_21)
# display(trips_21.take(10))

Row count: 2,495,635


In [0]:
# Your solution implementation to task 2.2 goes HERE
def t22_calc_profit(df):
    # input: output of task 2.1
    df=df.withColumn('unit_profitabilty',pf.when(df.trip_distance==0,0).otherwise(df.total_amount/df.trip_distance))
    return df

In [0]:
# execute task 2.2
trips_22 = t22_calc_profit(trips_21)

print_count(trips_22)
# display(trips_22.take(10))

Row count: 2,495,635


## Task 3: Rank zones by traffic, passenger volume and profitability

### 3.1 - Summarise interzonal travel

Build a graph data structure of zone-to-zone traffic, representing aggregated data about trips between any two zones. The graph will have one node for each zone and one edge connecting each pair of zones. In addition, edges contain aggregate information about all trips between those zones. 

For example, zones Z1 and Z2 are connected by *two* edges: edge Z1 --> Z2 carries aggregate data about all trips that originated in Z1 and ended in Z2, and edge Z2 --> Z2 carries aggregate data about all trips that originated in Z2 and ended in Z1.

The aggregate information of interzonal travel must include the following data:

- `average_unit_profit` - the average unit profitability (calculated as `mean(unit_profitabilty)`).
- `trips_count` -- the total number of recorded trips.
- `total_passengers` -- the total number of passenger across all trips (sum of `passenger_count`).

This graph can be represented as a new dataframe, with schema:

\[`PULocationID`, `DOLocationID`, `average_unit_profit`, `trips_count`, `total_passengers` \]

__hint__: the `groupby()` operator produces a `pyspark.sql.GroupedData` structure. You can then calculate multiple aggregations from this using `pyspark.sql.GroupedData.agg()`: 
- https://spark.apache.org/docs/3.2.0/api/python/reference/pyspark.pandas/api/pyspark.pandas.DataFrame.groupby.html
- https://spark.apache.org/docs/3.2.0/api/python/reference/api/pyspark.sql.GroupedData.agg.html

### Task 3.2 - Obtain top-10 zones

For each of the following measures, report the top-10 zones _using their plain names you dereferenced in the previous step, not the codes_. Note that this requires ranking the nodes in different orders. Specifically, you need to calculate the following further aggregations:

- the **total** number of trips originating from Z. This is simply the sum of `trips_count` over all outgoing edges for Z, i.e., edges of the form Z -> \*
- the **average** profitability of a zone. This is the average of all `average_unit_profit` over all *outgoing* edges from Z.
- The **total** passenger volume measured as the **sum** of `total_passengers` carried in trips that originate from Z

In [0]:

def t31_summarise_trips(df):
    # input: output of task 2.2
    df=df.groupby(df.PULocationID,df.DOLocationID).agg(pf.avg(df.unit_profitabilty).alias('average_unit_profit'),pf.count('*').alias('trips_count'),pf.sum(df.passenger_count).alias('total_passengers'))
    return df

In [0]:
# execute task 3.1
graph = t31_summarise_trips(trips_22)

print_count(graph)
# display(graph.take(10))

Row count: 13,700


In [0]:

def t32_summarise_zones_pairs(df, zones_df = zone_names):
    df=df.alias('a').join(zone_names.alias('b'),(df.PULocationID==zone_names.LocationID),'left').select('a.*','b.Zone')
    df=df.alias('a').groupby(df.Zone).agg(pf.sum(df.trips_count).alias('total_trips_count'),pf.sum(df.total_passengers).alias('total_passengers'),
                                          pf.avg(df.average_unit_profit).alias('average_unit_profit'))  
    return df

# Top 10 ranked zones by traffic (trip volume)
def t32_top10_trips(df_zones):
    # input: output of task 3.2
    df_zones=df_zones.sort(df_zones.total_trips_count.desc()).take(10)
    return df_zones

# Top 10 ranked zones by profit
def t32_top10_profit(df_zones):
    # input: output of task 3.2
    df_zones=df_zones.sort(df_zones.average_unit_profit.desc()).take(10)
    return df_zones

# Top 10 ranked zones by passenger volume
def t32_top10_passenger(df_zones):
    # input: output of task 3.2
    df_zones=df_zones.sort(df_zones.total_passengers.desc()).take(10)
    return df_zones

In [0]:
# execute task 3.2
zones = t32_summarise_zones_pairs(graph)

top10_trips     = t32_top10_trips(zones)
top10_profit    = t32_top10_profit(zones)
top10_passenger = t32_top10_passenger(zones)

In [0]:
# use 'display()' or return a pandas DataFrame for 'pretty' output
display(top10_trips)

Zone total_trips_count total_passengers average_unit_profit Upper East Side South 122366 170668.0 6.501500341406497 Upper East Side North 104192 144603.0 6.654370562642868 Penn Station/Madison Sq West 97209 137864.0 7.496291282263304 Midtown Center 94045 135530.0 7.191110506582673 Murray Hill 88873 124433.0 7.375015675416335 Midtown East 85651 120495.0 6.89901198344205 Lincoln Square East 81298 115223.0 7.060742273257432 Upper West Side South 75146 106473.0 6.485396949837054 Clinton East 74960 107811.0 7.484271876824437 Lenox Hill West 74693 104254.0 6.16758847663936

In [0]:
# use 'display()' return a pandas DataFrame for 'pretty' output
display(top10_profit)

Zone total_trips_count total_passengers average_unit_profit Bay Terrace/Fort Totten 49 5.0 311.9330683658483 Baisley Park 320 101.0 100.00592710440851 Prospect Park 74 64.0 90.58902850606826 Saint Michaels Cemetery/Woodside 29 29.0 66.14679218472433 Van Cortlandt Village 205 48.0 48.69059979610944 Grymes Hill/Clifton 2 2.0 38.65 Crotona Park 7 3.0 36.84437137879828 Flatbush/Ditmas Park 681 118.0 32.940319652777696 Eastchester 147 16.0 31.2081577880231 Flushing Meadows-Corona Park 53 61.0 29.53958083762334

In [0]:
# use 'display()' or return a pandas DataFrame for 'pretty' output
display(top10_passenger)

Zone total_trips_count total_passengers average_unit_profit Upper East Side South 122366 170668.0 6.501500341406497 Upper East Side North 104192 144603.0 6.654370562642868 Penn Station/Madison Sq West 97209 137864.0 7.496291282263304 Midtown Center 94045 135530.0 7.191110506582673 Murray Hill 88873 124433.0 7.375015675416335 Midtown East 85651 120495.0 6.89901198344205 Lincoln Square East 81298 115223.0 7.060742273257432 Clinton East 74960 107811.0 7.484271876824437 Upper West Side South 75146 106473.0 6.485396949837054 East Village 74243 106252.0 7.36438559455207

## Task 4 - Record the pipeline's execution time

Record the execution time of:

1. the whole pipeline
2. the whole pipeline except task 1.2

on the two tables below, for all dataset sizes: `'S'`, `'M'`, `'L'`, `'XL'`, `'XXL'`, and data formats: `parquet` and `delta`.

Analyse the resulting execution times and comment on the effect of dataset size, dataset format and task complexity (with and without task 1.2) on pipeline performance.

In [0]:
# after developing your solution, it may be convenient to combine all your functions in a single cell (at the start or end of the notebook)

# CHANGE the value of the following arguments to record the pipeline execution times for increasing dataset sizes
SIZE = 'XXL'
DATA_FORMAT = 'parquet'
WITH_TASK_12 = 0

# Load trips dataset
trips = init_trips(SIZE, DATA_FORMAT)


    Trips dataset loaded!
    ---
      Size: XXL
      Format: parquet
      Tables loaded: ['/FileStore/tables/taxi/parquet/tripdata_1_6_2019.parquet', '/FileStore/tables/taxi/parquet/tripdata_7_12_2019.parquet', '/FileStore/tables/taxi/parquet/tripdata_2020_21.parquet']
      Number of trips (dataset rows): 132,396,785
    


In [0]:
# run and record the resulting execution time shown by databricks (on the cell footer)

# IMPORTANT: this function calls all task functions in order of occurrence. For this code to run without errors, you have to load into memory all of the previous task-specific functions, even if you haven't implemented these yet.
pipeline(trips, with_task_12 = WITH_TASK_12)

Out[70]: [[Row(Zone='Upper East Side South', total_trips_count=5547013, total_passengers=8265045.0, average_unit_profit=5.730035925094978),
  Row(Zone='Upper East Side North', total_trips_count=5116488, total_passengers=7626015.0, average_unit_profit=5.270308317752348),
  Row(Zone='Midtown Center', total_trips_count=4874999, total_passengers=7456960.0, average_unit_profit=6.832096269925726),
  Row(Zone='Penn Station/Madison Sq West', total_trips_count=4366450, total_passengers=6523834.0, average_unit_profit=6.258419060591541),
  Row(Zone='Midtown East', total_trips_count=4307048, total_passengers=6472875.0, average_unit_profit=6.4566229774786),
  Row(Zone='Times Sq/Theatre District', total_trips_count=3844456, total_passengers=6100483.0, average_unit_profit=7.236831941928027),
  Row(Zone='Murray Hill', total_trips_count=3813421, total_passengers=5691669.0, average_unit_profit=6.786469187421671),
  Row(Zone='Clinton East', total_trips_count=3780351, total_passengers=5816084.0, average_u

_Table 1. Pipeline performance for `parquet` format._

| metric                      | S    | M    | L    | XL   | XXL  |
|-----------------------------|------|------|------|------|------|
| rows (M)                    |  2,898,033 |  15,571,166 |  41,953,716 |  90,443,069 |  132,396,785 |
| execution time   (w/o 1.2)  | 7.19s | 1.53min| 13.63min | 18.26min | 19.21min |
| execution time              | 15.21s | 3.14min | 30.63min | 40.62min | 41.69min |
| sec / 1M records (w/o 1.2)  | 2.48  | 5.89s  | 19.52s  | 12.11s  | 8.70s  |
| sec / 1M records            | 5.24  | 12.09s  | 43.80s  | 26.95s  | 18.89s  |

_Table 2. Pipeline performance for `delta` format._

| metric                      | S    | M    | L    | XL   | XXL  |
|-----------------------------|------|------|------|------|------|
| rows (M)                    |  2,898,033 |  15,571,166 |  41,953,716 |  90,443,069 |  132,396,785 |
| execution time   (w/o 1.2)(sec)  | 6.44s | 7.58s | 8.74s | 18.60s | 39.44s |
| execution time(sec)              | 12.20s | 18.05s | 18.79s | 39.45s | 51.89s |
| sec / 1M records (w/o 1.2)  | 2.22s  | 0.486s  | 0.20s  | 0.20s  | 0.29s  |
| sec / 1M records            | 4.20s  | 1.15s  | 0.44s  | 0.44s  | 0.39s  |